### DISCUSSION TITLE
The idea of Backend Struct

### DISCUSSION THREAD

**Discussion by HosMercury** [maxcountryman/axum-login#186] [Ideas]
I think this struct makes things very overwhelming 

you implement AuthBackend for Backend 
It should be Auth for User and Authz for User 

So we could impl User more like User.register  , User.is_active .. etc ?

the crate is amazing except that deformation !

what about removing this Bacend Struct at all and just receive the client and db pool when needed by the user of the crate ??
wdyt ?

**Comment by maxcountryman** [comment]
You may implement `AuthnBackend` and `AuthzBackend` for any arbitrary type you like (where the latter requires `AuthnBackend` to also be implemented for the type). The name of that type may be anything you like and the examples merely use `Backend`, as, well, an example.

The important thing to understand is that these traits enable _arbitrary_ authentication backends which, in a sense, produce users (more specifically authenticated users, see below). What that means is `axum-login` can support virtually any kind of authentication, provided at least `AuthnBackend` can be implemented for it.

As far as the user type, this is also _arbitrary_: you must implement `AuthUser` for your user type, but you're free to implement any other methods you might like, including things like `is_active` (it's up to your application to decide what this semantic is and that's an intentional design decision).

Putting it all together, `axum-login` works by providing a generic interface for applications to define how authenticated (and optionally authorized) users can be produced from some backend (a database, OAuth, LDAP, an IdP, etc).

As I mentioned in the other thread, the concept of an auth backend is the primary basis of this crate. Given that, we won't be renaming the traits or removing them from the crate.

**Reply by HosMercury** [reply]
May be bc I'm not professional enough, but it was an isea .
